File: lib/PList.cpp
#include <stdio.h>

class Object {

};

class StringItem : public Object {
public:
	const char * str;
	StringItem(const char *as = "")
	{
  		str = as;
  	}
};

class IntegerItem : public Object {
public:
	int i;
	IntegerItem(int n = 0) : i(n) {}
};

class DoubleItem : public Object {
public:
	double i;
	DoubleItem(double n = 0.0) : i(n) {}
};

class List {
private:
	class ListNode {
	private:
		Object obj; 
		ListNode *next;
	public:
		ListNode(Object a) 
			{ obj = a; next=NULL; }
		ListNode* getNext() { return next; }
		void setNext(ListNode *n) { next = n; }
		Object getItem() { return obj; }
	};
	ListNode *head;
	ListNode *tail;
public:
	List()
	{
		head = NULL;
		tail = NULL;
	}
	void push_back(Object a)
	{
		ListNode *node = new ListNode(a);
		if (head==NULL) {  
			head = node;
			tail = node;
		} else {
			tail->setNext(node); 
			tail = node;         
		}
	}  
	bool remove_front(Object &a)
	{
		if (!empty()) 
		{
			ListNode *tmp = head->getNext();
			delete head; 
			head = tmp;  
			if (tmp==NULL) 
				tail = NULL;
			return true;
		}
		return false; 
	}   
	bool empty()
	{
		return head==NULL;
	}
	Object get(int n)
	{
		int i = 0;
		Object o;
		ListNode* currNode = head;
		
		while (currNode != NULL)
		{
			if (i == n)
				return currNode->getItem();		
			i++;
			currNode = currNode->getNext();
		}
		
		return o;
	}
	int length()
	{
		int i = 0;
		ListNode* currNode = head;
		
		while (currNode != NULL)
		{
			++i;
			currNode = currNode->getNext();
		}
		return i;
	}
	~List()
	{
		Object t;
		while(!empty()) 
			remove_front(t);   
	}
};

File: python/PList.py
from __future__ import annotations

class Object:
    def __init__(self):
        pass
class StringItem(Object):
    def __init__(self, c=""):
        s = c
class IntegerItem(Object):
    def __init__(self, n = 0):
        i = int(n)
class DoubleItem(Object):
    def __init__(self, n = 0.0):
        i = float(n)
class PList:
    def __init__(self):
        self.head = None
        self.tail = None
    def push_back(self, a: Object) -> None:
        node = ListNode(a)
        if self.head == None:
            self.head = node
            self.tail = node
        else:
            self.tail.setNext(node)
            self.tail = node
    def remove_front(self, a: Object) -> bool:
        if not self.empty():
            tmp = self.head.getNext()
            self.head = tmp
            if tmp == None:
                self.tail = None
            return True
        return False
    def empty(self) -> bool:
        return self.head == None
    def get(self, n: int) -> Object:
        i = 0
        o = Object()
        currNode = self.head
        while currNode != None:
            if i == n:
                return currNode.getItem()
            i += 1
            currNode = currNode.getNext()
        return i
    def __del__(self):
        t = Object()
        while not self.empty():
            self.remove_front(t)
class PListNode(PList):
    def __init__(self, a: Object):
        self.obj = a
        self.nxt = None
    def getNext(self) -> PListNode:
        return self.next
    def setNext(self, n: PListNode) -> None:
        self.nxt = n
    def getItem(self) -> Object:
        return self.obj
